### DISCUSSION TITLE
[BUG] PyArmor Obfuscated Scripts Not working on AWS Lambda (with Custom Container Image)

### DISCUSSION THREAD

**Discussion by Vishal-Shunyeka** [dashingsoft/pyarmor#1151] [Q&A] [answered]
When running the Obfuscated Scripts on AWS Lambda the Lambda Function exits with this error `Error: Runtime exited with error: exit status 1`. The image works fine on local. This could be an issue with pyarmor not working fine with lambda interface and If not, what could be the solution to making this work?

The DockerFile and python files to recreate the issue:

DockerFile:
```
### create custom image function which contains library
FROM public.ecr.aws/lambda/python:3.9

ENV LANG=en_US.UTF-8
ENV TZ=:/etc/localtime
ENV PATH=/var/lang/bin:/usr/local/bin:/usr/bin/:/bin:/opt/bin
ENV LD_LIBRARY_PATH=/var/lang/lib:/lib64:/usr/lib64:/var/runtime:/var/runtime/lib:/var/task:/var/task/lib:/opt/lib:/usr/local/lib
ENV LAMBDA_TASK_ROOT=/var/task
ENV LAMBDA_RUNTIME_DIR=/var/runtime
ENV PYARMOR_CLI=7

RUN yum install -y git unzip gcc-c++ make
RUN pip3 install pyarmor==7.7.4

# Copy allfile to docker image
COPY ./ ${LAMBDA_TASK_ROOT}

# Install the function's dependencies using file requirements.txt in your project folder.
RUN pip3 install -r requirements.txt --target "${LAMBDA_TASK_ROOT}"

# Obfuscate Services
RUN pyarmor obfuscate obfuscated.py --exclude ./venv/ --restrict=0
RUN rm -f obfuscated.py
RUN mv ./dist/obfuscated.py .
RUN mv ./dist/pytransform ./

RUN chmod +x obfuscated.py
RUN chmod +x lambda_code.py
# Set the CMD to your handler (could also be done as a parameter override outside of the Dockerfile)
CMD [ "lambda_code.lambda_handler" ]
```

obfuscated.py
```
#! /bin/env python
def obfuscated_function():
    print("It succeeded.")
    return "Success"

```

lambda_code.py
```
#! /bin/env python
import json
from obfuscated import obfuscated_function

print('Loading function')


def lambda_handler(event, context):
    print("Received event: " + json.dumps(event, indent=2))
    return obfuscated_function()

```

Edit: Updated the files as per suggestions, still not working.

**Comment by jondy** [comment]
I'm not sure, but please add one line something like `#! /bin/env python` and `chmod +x` for obfuscated scripts. 

Please also do a simple test, obfuscate a hello world script, and run it with python, check the obfuscated scripts or not.

**Comment by Vishal-Shunyeka** [comment]
It works on local, but not when the image is uploaded to lambda, did the changes and it still does not work on lambda.
Error :
```
{
  "errorType": "Runtime.ExitError",
  "errorMessage": "RequestId: 66ec6b5e-75e2-4723-aaf2-0f01104c4f91 Error: Runtime exited with error: exit status 1"
}
```

Updated the file contents in Issue to reflect new files.

**Comment by joonis** [comment, marked as answer]
AWS Lambda does not work with the anti-debug feature.

**Reply by jonathanslenders** [reply]
@joonis : Can you elaborate on this? What would be needed on the AWS side for this to be fixed?

Could it be a case where Pyarmor creates shared objects that dynamically link certain (system) shared libraries, which are possibly not present by default?

**Reply by joonis** [reply]
I don't know the exact reason, but Pyarmor 7 worked without anti-debug features. Have not tested Pyarmor 8.

**Comment by LutzFassl** [comment]
I have the same problem. It works when I test the lambda docker image locally (When using RIE as per (https://docs.aws.amazon.com/lambda/latest/dg/images-test.html).

**But** when I try within the actual lambda function it always has the same problem like OP said: 
**Error: Runtime exited with error: exit status 1 Runtime.ExitError**

As soon as the first obfuscated module is imported.

**Comment by jondy** [comment]
If anti-debug is root cause, Pyarmor 8 should work without `--enable-jit` and `--enable-bcc`

**Comment by jondy** [comment]
I thought of something, `pyarmor_runtime.so` need `libpythonX.Y.so` and a few dependent dynamic libraries, check it by `objdump -T /path/to/pyarmor_runtime.so`.

I don't know whether these dependent libraries are  installed in the AWS Lambda or not.
